### ISSUE TITLE
Silent failure on regexp stack overflow on wasi

### ISSUE THREAD

**Issue by dkegel-fastly** [tinygo-org/tinygo#3258]
A large wasi app compiled fine but inexplicably started failing its smoke test when a small change was made to the source tree.

The failure had nothing obvious to do with the change; the connection only became apparent when I bisected the change.
After a little reduction, the failure was no longer silent, but involved a wasm trap.

Further bisection got the problem down to a single possibly pathological regexp, which triggers a trap all by itself.

tinygo version reports
```
tinygo version 0.26.0-dev-f4395147 darwin/amd64 (using go version go1.18.7 and LLVM version 14.0.0)
```

It would be nice if that case were caught more reliably, and weren't silent in the larger test case.  Alas, I do not have a minimal reproducer for the silent part, but here is a small reproducer for the kind of trap that somehow because silent in the larger program:

```
$ cat foo.go
package main
import "regexp"
var r = regexp.MustCompile(`(x[a-z]*[ ]{0,10}=[ ]{0,10}y[ ]{0,10}\(.*\))`)
func main() {}
$ go run foo.go
$ tinygo run foo.go
$ tinygo run -target wasi foo.go
panic: runtime error: index out of range

Caused by:
    0: failed to invoke command default
    1: wasm trap: wasm `unreachable` instruction executed
       wasm backtrace:
           0:  0x9f0 - <unknown>!runtime.runtimePanic
           1:  0x442 - <unknown>!runtime.lookupPanic
           2: 0x3d77 - <unknown>!(regexp/syntax.patchList).patch
           3: 0x44f8 - <unknown>!(*regexp/syntax.compiler).cat
           4: 0x500f - <unknown>!(*regexp/syntax.compiler).compile
           5: 0x50d6 - <unknown>!(*regexp/syntax.compiler).compile
           6: 0x4fb4 - <unknown>!(*regexp/syntax.compiler).compile
           7: 0x50d6 - <unknown>!(*regexp/syntax.compiler).compile
           8: 0x4fb4 - <unknown>!(*regexp/syntax.compiler).compile
           9: 0x50d6 - <unknown>!(*regexp/syntax.compiler).compile
          10: 0x4fb4 - <unknown>!(*regexp/syntax.compiler).compile
          11: 0x50d6 - <unknown>!(*regexp/syntax.compiler).compile
          12: 0x4fb4 - <unknown>!(*regexp/syntax.compiler).compile
          13: 0x552c - <unknown>!(*regexp/syntax.compiler).compile
          14: 0x1e201 - <unknown>!runtime.run$1
          15: 0x1da74 - <unknown>!runtime.run$1$gowrapper
          16: 0x202ee - <unknown>!tinygo_launch
          17: 0x1d916 - <unknown>!_start
       note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable to may show more debugging information
       
error: failed to run compiled binary /var/folders/v0/0k9hwftd29b18z9z7s2g2sfc0000gn/T/tinygo1536841757/main: exit status 134
```

**Comment by dkegel-fastly** [comment]
Moving the regexp compile out of init time still fails on wasi, this time with a stack overflow:

```
package main
import (
	"bufio"
	"os"
	"regexp"
)
func main() {
	var r = regexp.MustCompile(`(x[a-z]*[ ]{0,10}=[ ]{0,10}y[ ]{0,10}\(.*\))`)
	reader := bufio.NewReader(os.Stdin)
	text, _ := reader.ReadString('\n')
	println(r.MatchString(text))
}
$ echo hi | tinygo run -target wasi foo.go
panic: runtime error: stack overflow
Error: failed to run main module `/var/folders/v0/0k9hwftd29b18z9z7s2g2sfc0000gn/T/tinygo3235268697/main`

Caused by:
    0: failed to invoke command default
    1: wasm trap: wasm `unreachable` instruction executed
       wasm backtrace:
           0: 0x47e8 - <unknown>!runtime.runtimePanic
           1: 0x46e3 - <unknown>!internal/task.Pause
           2: 0x8b27 - <unknown>!runtime.deadlock
           3: 0x11dfb - <unknown>!runtime.run$1$gowrapper
           4: 0x3c268 - <unknown>!tinygo_launch
           5: 0x11c88 - <unknown>!_start
```

**Comment by dkegel-fastly** [comment]
Another clue to the silent failure: the initial trap in the larger program after I whittled it down a little looks like this:
```
2022-10-29T16:13:53.025538Z ERROR request{id=1}: WebAssembly trapped: wasm trap: uninitialized element
wasm backtrace:
    0: 0x18c279 - <unknown>!tinygo_launch
    1: 0x1fdbb - (*internal/task.Task).Resume
                    at /Users/dkegel/src/tinygo/src/internal/task/task_asyncify.go:110:17
              - runtime.scheduler
                    at /Users/dkegel/src/tinygo/src/runtime/scheduler.go:236:11
              - runtime.run
                    at /Users/dkegel/src/tinygo/src/runtime/scheduler_any.go:29:11
              - _start
                    at /Users/dkegel/src/tinygo/src/runtime/runtime_wasm_wasi.go:22:5
```

**Comment by dkegel-fastly** [comment]
Given https://github.com/tinygo-org/tinygo/commit/d435fc868b9caef76b7a9b30a9c33adc3a79cab4 I should repeat this test with the tip of dev...

**Comment by dkegel-fastly** [comment]
Happily, my app doesn't actually need the silly regexp.
But for others running into this, one possible kludge is to replace quantifiers like "{1,32}" with simpler ones like "+".

**Comment by dgryski** [comment]
Nice debugging! I wonder if it's a memory corruption issue? I've had issues that led to silent corruption when the stack overflowed.

**Comment by aykevl** [comment]
Interesting! I can reproduce this on the dev branch.

Unfortunately, it seems to go away with `-gc=leaking` or `-scheduler=none` or `-opt=1`. Which makes debugging harder. In any case, this looks a lot like a memory corruption.

**Comment by dkegel-fastly** [comment]
FWIW my big app uses -gc=leaking and -opt=2 and it was still affected...?

**Comment by aykevl** [comment]
Well, that rules out an incorrectly freed object, then.

**Comment by aykevl** [comment]
...could it simply be an actual stack overflow? Because I can't reproduce this anymore with `-stack-size=32kb`.

**Comment by dkegel-fastly** [comment]
That's what I suspect.  I hadn't thought to try fiddling with that parameter, even though I asked for it :-)

So the bug is "stack overflows are sometimes silent", I guess, and/or "default stack on wasi is too small"?

**Comment by aykevl** [comment]
> stack overflows are sometimes silent

Yeah, that's true on all platforms TinyGo supports at the moment. Not sure whether this is something we want to fix (it would be great, but I'm not sure whether there is a practical and performant way to do it).

> default stack on wasi is too small

That's subjective :) larger stacks also currently mean more data to scan each GC cycle. (This should be fixable though with a precise GC).

**Comment by dkegel-fastly** [comment]
Maybe only do it on test executables...?

**Comment by aykevl** [comment]
We could, but I'm not so sure that's a good idea. It could lead to rather surprising issues where a test executable works but the regular code doesn't.

**Comment by dkegel-fastly** [comment]
Was any part of my problem specific to build time interpretation?  It feels like the silent failure part might have been, but I have no idea... If so, maybe build time interpretation could get a stack overflow check and/or a bigger stack.

**Comment by dgryski** [comment]
The 'standard' way to check for stack overflows is with canary bytes that are overwritten and then checked at function return time.

**Comment by aykevl** [comment]
> Was any part of my problem specific to build time interpretation?

I doubt it.

> If so, maybe build time interpretation could get a stack overflow check and/or a bigger stack.

That's not a thing. The interp pass is regular Go code, so it uses as much stack space as the host provides (by default 1GB).

> The 'standard' way to check for stack overflows is with canary bytes that are overwritten and then checked at function return time.

Perhaps for C (which allows arbitrarily sized stack allocations), but there are more efficient ways to do this in TinyGo because I made sure all functions have a fixed frame size. Basically, this can be done by checking the stack pointer right after incrementing it at the function entry. That's possible, but requires some cooperation with the LLVM WebAssembly backend to be efficient.

The alternative method with canary bytes seems far more involved, and requires the same amount of backend cooperation.

**Comment by dkegel-fastly** [comment]
See also https://www.theregister.com/2021/11/04/webassembly_stack_canaries/

**Comment by aykevl** [comment]
@dkegel-fastly there are many problems with that report, to be honest it mostly feels like FUD to me. I think the writers of that paper don't really understand the security model of WebAssembly. While it's indeed harder to work around C language deficiencies (such as memory unsafety) in the compiler, this is mostly besides the point as WebAssembly itself is sandboxed. (Possibly a hot take here)

...also, it's The Register so I'm not exactly expecting high quality journalism there.

**Comment by dkegel-fastly** [comment]
Tinygo's undefined stack overflow behavior is fine as is if you're used to low level development in C or assembly, and have good tests.

It would be a bonus if it had solid stack overflow behavior at least when running on bigger targets (including wasm), but there are bigger fish to fry...

**Comment by aykevl** [comment]
Yeah, it would be really useful to have. Probably as a late LLVM machine code pass. It could also be useful for Rust or for RTOSes that want to provide more safety on bare metal systems. For Linux/macOS/Windows, this can be more easily solved with guard pages (which we don't support either at the moment).

**Comment by dgryski** [comment]
I wonder if we could have the function prologue check if we're about to overflow the stack and panic with a custom "stack oom" error?

**Comment by aykevl** [comment]
@dgryski yeah that's what I mean, it has to be part of the LLVM backends. LLVM IR is too early in the compilation process. Or we'd need to patch binaries with a new pass like Asyncify does for example.
